### PR TITLE
[ao] Added README for ModelReport functionality

### PR DESCRIPTION
**Pull request by vspenubarthi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81369
* #81382

Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/fx/_model_report/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+There is a tutorial in the works that will walk through a full usage of the ModelReport API.
+This tutorial will show the ModelReport API being used on toy model in both an Fx Graph Mode workflow and an alterative workflow with just a traceable model.
+This README will be updated with a link to the tutorial upon completion of the tutorial.
+
+## Implementing Your Own Detector
+
+The main way to add functionality to the ModelReport API is to add more Detectors.
+Detectors each have a specific focus in terms of the type of information they collect information on.
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/README.md:66]: type of information they collect?

File: torch/ao/quantization/fx/_model_report/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+For example, the `DynamicStaticDetector` figures out whether Dynamic or Static Quantization is appropriate for different layers.
+Meanwhile, the `InputWeightEqualizationDetector` determines whether Input-Weight Equalization should be applied for each layer.
+
+
+### Requirements to Implement Detector
+All Detectors inherit from the `DetectorBase` class, and all of them (including any custom detectors you create) will need to implement 3 methods:
+- `determine_observer_insert_points(self, model)` -> `Dict`: determines which observers you want into insert into a model to gather statistics and where in the model. All of them return a dictionary mapping unique observer fully qualified names (fqns), which is where we want to insert them, to a dictionary of location and argument information in the format:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/README.md:73]:
```suggestion
- `determine_observer_insert_points(self, model)` -> `Dict`: determines which observers you want to insert into a model to gather statistics and where in the model. All of them return a dictionary mapping unique observer fully qualified names (fqns), which is where we want to insert them, to a dictionary of location and argument information in the format:
```

File: torch/ao/quantization/fx/_model_report/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+```python
+return_dict = {
+    "[unique_observer_fqn_of_insert_location]" :
+    {
+        "target_node" -> the node we are trying to observe with this observer (torch.fx.node.Node),
+        "insert_observer" -> the observer we wish to insert (ObserverBase),
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/README.md:80]: I feel `insert_observer` is not very intuitive, maybe rename this to `observer_to_insert` or something similar?

File: torch/ao/quantization/fx/_model_report/README.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        "insert_post" -> True if this is meant to be a post-observer for target_node, False if pre-observer,
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/README.md:81]: typically bool will starts with `is_`, maybe `is_post_observer`?

File: torch/ao/quantization/fx/_model_report/README.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+- `__init__(self, desired_report_detectors: Set[DetectorBase])` constructor that takes in instances of desired detectors and stores them.
+This is so that we can keep track of where we want to insert observers on a detector by detector basis and also keep track of which detectors to generate reports for.
+- `prepare_detailed_calibration(self, model: GraphModule)` &rarr; `GraphModule` takes in a GraphModule and inserts observers into the locations specified by each detector.
+It then returns the passed in GraphModule with the detectors inserted into both the regular module structure as well as the node structure.
+- `generate_model_report(self, fx_model: GraphModule, **kwargs)` &rarr; `Dict[str, Tuple[str, Dict]]` takes in a GraphModule that has been callibrated and generates, for each detector the ModelReport instance was initialized with:
+  - A string-based report that is easily digestable and actionable explaining the data collected by relavent observers for that detector
+  - A dictionary containing statistics collected by the relavent observers and values calculated by the detector for futher analysis or plotting
+
+
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/README.md:17]: should you talk about `ModelReporterObserver` and it's relationship with detectors as well

File: torch/ao/quantization/fx/_model_report/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+## Implementing Your Own Detector
+
+The main way to add functionality to the ModelReport API is to add more Detectors.
+Detectors each have a specific focus in terms of the type of information they collect information on.
+For example, the `DynamicStaticDetector` figures out whether Dynamic or Static Quantization is appropriate for different layers.
+Meanwhile, the `InputWeightEqualizationDetector` determines whether Input-Weight Equalization should be applied for each layer.
+
+
+### Requirements to Implement Detector
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/_model_report/README.md:71]: I feel the design for detector should be close to model reporter design overview. and also we can talk about high level flow and usage before talking about design details

**Review by jerryzh168 — APPROVED** [review]
looks good overall, had a few comments inline

File: torch/ao/quantization/fx/_model_report/README.md (hunk 3 above)
Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/README.md:80]: Addressed in new downstream commit here: https://github.com/pytorch/pytorch/pull/81382

File: torch/ao/quantization/fx/_model_report/README.md (hunk 4 above)
Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/README.md:81]: Addressed in new downstream commit here: https://github.com/pytorch/pytorch/pull/81382

**Pull request closed without merging**